File: Server/handlers/funcs/udp/udp.go
package udpfuncs

import (
	"bytes"
	"fmt"

	"atxmedia.us/multiassembly/handlers/bit"
	"atxmedia.us/multiassembly/player"
	"atxmedia.us/multiassembly/util"
)

var funcs map[string]func(*player.Player, *bytes.Buffer) bool = make(map[string]func(*player.Player, *bytes.Buffer) bool)

func init() {
	funcs["PTUP"] = playerTransformUpdatePosition
	funcs["PTUR"] = playerTransformUpdateRotation
}

func Run(uuid string, fcfi string, buf *bytes.Buffer) bool {
	fn, ok := funcs[fcfi]
	if !ok {
		fmt.Println("Invalid FCFI for UDP function")
		return false
	}
	p := player.GetByID(uuid)
	if p == nil {
		return false
	}
	return fn(p, buf)
}

func playerTransformUpdatePosition(p *player.Player, buf *bytes.Buffer) bool {
	x, ok := bit.ReadFloat64(buf)
	if !ok {
		return false
	}
	y, ok := bit.ReadFloat64(buf)
	if !ok {
		return false
	}
	z, ok := bit.ReadFloat64(buf)
	if !ok {
		return false
	}
	p.Position = util.Vector3{X: x, Y: y, Z: z}
	return true
}
func playerTransformUpdateRotation(p *player.Player, buf *bytes.Buffer) bool {
	x, ok := bit.ReadFloat64(buf)
	if !ok {
		return false
	}
	y, ok := bit.ReadFloat64(buf)
	if !ok {
		return false
	}
	z, ok := bit.ReadFloat64(buf)
	if !ok {
		return false
	}
	p.Rotation = util.Vector3{X: x, Y: y, Z: z}
	return true
}

File: MultiAssembly/Handlers/UDP.cs
using System;
using System.Collections.Generic;
using System.IO;
using UnityEngine;
using System.Text;

namespace MultiAssembly.Handlers
{
    internal class UDP
    {
        public static Dictionary<string, Action<MemoryStream>> functions = new Dictionary<string, Action<MemoryStream>>();

        static UDP()
        {
            functions.Add("PTUP", playerTransformUpdatePosition);
            functions.Add("PTUR", playerTransformUpdateRotation);
        }

        public static void Run(string fcfi, MemoryStream stream)
        {
            if (!functions.TryGetValue(fcfi, out Action<MemoryStream> func))
            {
                throw new HandlerNotFoundException(fcfi);
            }
            func(stream);
        }

        private static void playerTransformUpdatePosition(MemoryStream stream)
        {
            string uuid = Bit.ReadString(stream, UUID.UUIDLength);
            float x = (float)Bit.ReadDouble(stream);
            float y = (float)Bit.ReadDouble(stream);
            float z = (float)Bit.ReadDouble(stream);
            Player? p = Player.Find(uuid);
            //Check for null transform because of Unity cleanup
            if (p == null || p.GetGameObject().transform == null)
            {
                return;
                //throw new PlayerNotFoundException(uuid);
            }
            p.GetGameObject().transform.position = new Vector3(x, y, z);
        }
        private static void playerTransformUpdateRotation(MemoryStream stream)
        {
            string uuid = Bit.ReadString(stream, UUID.UUIDLength);
            float x = (float)Bit.ReadDouble(stream);
            float y = (float)Bit.ReadDouble(stream);
            float z = (float)Bit.ReadDouble(stream);
            Player? p = Player.Find(uuid);
            if (p == null || p.GetGameObject().transform == null)
            {
                return;
                //throw new PlayerNotFoundException(uuid);
            }
            p.GetGameObject().transform.eulerAngles = new Vector3(x, y, z);
        }
    }
}
